bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerDescriptor.java

```diff
@@ -47,7 +47,7 @@ static LedgerDescriptor create(byte[] masterKey,
 
     static LedgerDescriptor createReadOnly(long ledgerId,
                                            LedgerStorage ledgerStorage)
-            throws IOException, Bookie.NoLedgerException {
+            throws IOException {
```

Comment:
Actually IOException is superclass of Bookie.NoLedgerException so even though method throws Bookie.NoLedgerException it will be wrapped into IOException anyways. In other words this method will only throw IOException.

Reply:
Well, the code will throw `Bookie.NoLedgerException` , which extend from IOExcpetion, however depends on the client code on how to catch this exception.

For example, if the client has code like:

try {

} catch (Bookie.NoLedgerException be) {
  // Do handle on No ledger exception
} catch (IOException ioEx) {
  // Do handle to IOException
} catch (Exception ex) {
  // Handle the rest 
}

in the case above, if the exception is `Bookie.NoLedgerException` , then it can be specifically handle the exception about no ledger.

Reply:
Yes. So basically if createReadOnly declares that it throws IOException and Bookie.NoLedgerException, essentially it means, It declares that it can throw IOException.  Since Bookie.NoLedgerException is subclass of IOException createReadOnly is well within its rights to throw Bookie.NoLedgerException.   And at the client code exception chaining will work exactly as you showed.

Reply:
I quickly wrote this tiny code. `class TestException extends Exception {
    TestException(String st) {
        super(st);
    }
}

public class TestClass {
    void testMethod(int i) throws Exception {
        System.out.println("Value of I = " + i);
        if (i % 2 == 0) {
            throw new Exception("except");
        } else {
            throw new TestException("test except");
        }
    }

    public static void main(String [] args) {
        TestClass t = new TestClass();
        try {
            t.testMethod(0);
           // t.testMethod(1);
        } catch (TestException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}`

With 0. Output is: 2 actionable tasks: 2 executed
java.lang.Exception: except
	at TestClass.testMethod(TestClass.java:13)
	at TestClass.main(TestClass.java:22)

With 1. 

BUILD SUCCESSFUL in 222ms
2 actionable tasks: 2 executed
TestException: test except
	at TestClass.testMethod(TestClass.java:15)
	at TestClass.main(TestClass.java:23)

Reply:
Yes, the `throws` clause just used as indicator of what kind of checked exceptions can be thrown by the method. 

Since the `Bookie.NoLedgerException` is also `IOException`, then by just using `throw Bookie.NoLedgerException` in the method declaration, the client code can still handle both  `Bookie.NoLedgerException` and `IOException` in the catch clauses. However, by indicating `throws IOException, Bookie.NoLedgerException` in the method signature, it gives the client or user code to know what that there are different kind of Exception can be thrown by the method call to give chance to handle them separately/ properly. If the method just have `throw  Bookie.NoLedgerException`, the client may not know that the method can also throw `IOException`. 

So, it is more of programming style of giving hints for caller to know what to expect from calling this method.

Reply:
Yeah. Makes sense.